Implementation of Consent Manager 2.0 + fixes after reviews

simhue opened the pull request.

simhue committed:
Add com.fasterxml.jackson.datatype.jsr310.JavaTimeModule for java.time.Instant serialization

simhue committed:
Angular CLI update for packages - @angular/core@9, @angular/cli@9

simhue committed:
@angular/cli migration - workspace-version-9
Angular Workspace migration. Update an Angular CLI workspace to version 9.

simhue committed:
remove obsolete files

simhue committed:
upgrade to Angular 18

simhue committed:
#3: first prototype of consent manager 2.0

simhue committed:
Merge branch 'master' into Ugrade-to-Angular-18

simhue committed:
fix errors that occurred due to upgrade to angular 18

simhue committed:
add lombok dependency

simhue committed:
https://github.com/aktin/dwh-admin/issues/3:
- optimized build inludes/excludes
- refactoring
- add new classes for master data and encounters
- apply new logo and icons

simhue committed:
remove obsolete file

simhue committed:
https://github.com/aktin/dwh-admin/issues/3:
- add comments
- refactoring
- implement batch registration

simhue committed:
https://github.com/aktin/dwh-admin/issues/3:
fixes after first review https://github.com/aktin/dwh-admin/issues/3#issuecomment-2653315180

simhue committed:
33: reworked modal/dialog logic

simhue committed:
#49: - let error interceptor handle http errors
- add case for http 304 code in error interceptor
- refactor request polling

simhue committed:
Merge branch 'develop' into consent-manager-20

simhue committed:
feat: #67: add custom theme

simhue committed:
feat: #67: add semantic sources to gitignore

simhue committed:
feat: #67: make fomantic ui build process more ci/cd friendly

simhue committed:
feat: #67: force add site.variables since re-inclusion in gitignore was not working

simhue committed:
feat: #67: remove obsolete files

simhue committed:
feat: #67: change primary color

simhue committed:
Merge branch 'theming' into cm2-with-theming

simhue committed:
fix: fix request result download

simhue committed:
fix: remove git merge tags

simhue committed:
feat: #67: unify width of pages

simhue committed:
change options

simhue committed:
#67: move fomantic ui to dev dependencies

simhue committed:
pass error to console as object

simhue committed:
#67: remove obsolete attributes

simhue committed:
fix: remove git diff notation

simhue committed:
fix: #33: refined modal logic and styling

simhue committed:
remove obsolete reference

simhue committed:
fix: #3: change wording

simhue committed:
fix: #3: add missing annotation to parameters

simhue committed:
fix: #3: adapt theme color coding for Multi-registration table

simhue committed:
fix: #3: add additional text that SIC is not required when not automatically generated

simhue committed:
Merge branch 'develop' into feat/consent-manager-20

Signed-off-by: Simon H. <huening.simon@gmail.com>

simhue committed:
#74: https://github.com/aktin/dwh-admin/issues/74#issuecomment-3425842268: disallow white spaces

simhue committed:
#74: https://github.com/aktin/dwh-admin/issues/74#issuecomment-3425842268: only allow pasting of excel data structure if no input element is selected

simhue committed:
Merge branch 'develop' into feat/consent-manager-20

Signed-off-by: Simon H. <huening.simon@gmail.com>

akomii committed:
updated aktin internal dependencies to current versions

akomii commented on admin-gui/src/main/webapp/semantic/src/site/globals/site.variables:
Duplicate with #73  ?

simhue replied:
You are right, I merged #73 branch into this one with commit 4ae3473

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/PatientEntriesRequestDTO.java:
```diff
@@ -0,0 +1,20 @@
+	// lists of sics and extensions, each element of one list corresponds to the element in the other list at the same index
+	// if a sic is going to be generated, the value of the respective index should be null
+	public List<String> sics;
+	public List<String> extensions;
```
Why are you using here parallel lists? This is very fragile. What was the reason not to use a map?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+        PatientEntry pat = study.getPatientByID(ref, root, ext);
+        if (pat != null) {
+            log.log(Level.WARNING, "Cannot create entry, PatientEntry already exists.");
+            return Response.status(Status.CONFLICT).entity("Patient*in existiert bereits").build();
```
Do not sent human-readable error messages. These are endpoints. A machine interacts with them (frontend). The translation to human-readable should be done in Frontend to UI

akomii replied:
Instead of Response you could also return:

throw new ClientErrorException(Status.CONFLICT);

This is a preformatted machine-readable status message

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+        pat = study.getPatientBySIC(entry.sic);
+        if (pat != null) {
+            log.log(Level.WARNING, "Cannot create entry, SIC already exists.");
+            return Response.status(Status.CONFLICT).entity("Studien-ID existiert bereits").build();
```
See above

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+        PatientEntry oldEntry = study.getPatientByID(ref, root, ext);
+        if (oldEntry == null) {
+            log.log(Level.WARNING, "Cannot update entry, entry does not exist");
+            return Response.status(Status.BAD_REQUEST).entity("Patient*in nicht gefunden").build();
```
Here too

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+        Study study = this.getStudy(id);
+        PatientEntry pat = study.getPatientByID(ref, root, ext);
+        if (pat == null) {
+            return Response.status(Status.BAD_REQUEST).entity("Patient*in nicht gefunden").build();
```
Here too

akomii replied:
Why even BAD_REQUEST?? Why not NOT_FOUND?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+    }
+
+    /**
+     * Gets an entry by the specified study id, reference type, root and extension parameters.
```
Not what the method does

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/JavaTimeMapperProvider.java:
Probably not necessary (fixed by downgrading WildFly)

simhue replied:
I needed this to properly seriallize Instant datatypes

akomii replied:
Still, please check if removing this class has an affect of the rest of the dwh

simhue replied:
The rest of dwh seems to work without the class, but Instant is also used in other packages, and I saw Raphael's comment in ReportEntry.java: 
´´´
// TODO: Instant instead Date (fix problem: receiving nano and epochSecond when using Instant)
public Date created;
public Date data;
public Date start;
public Date end;
´´´

akomii replied:
Classes/Depdencies should only be added if the affect the current scope. If the app works without JavaTimeMapperProvider, it should be removed

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+    @GET
+    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
+    public PatientEntry getEntryBySic(@PathParam("studyId") String id, @PathParam("sic") String sic) throws IOException {
+        Study study = this.getStudy(id);
```
What if study is null?

simhue replied:
getStudy throws a NotFound exception, resulting in a 404

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+            return Response.status(Status.BAD_REQUEST).entity("Patient*in nicht gefunden").build();
+        }
+        PatientEntry newEntry = study.getPatientByID(ref, root, ext);
+        newEntry.setComment(entry.comment);
```
Modifying an object should be done in the service, not the endpoint. The endpoint only receives commands which are forwarded to the service. Only for next time. Keep it as is now

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+            log.log(Level.WARNING, "Cannot update entry, entry does not exist");
+            return Response.status(Status.BAD_REQUEST).entity("Patient*in nicht gefunden").build();
+        }
+        PatientEntry newEntry = study.getPatientByID(ref, root, ext);
```
What?? So newEntry = oldEntry? Why not just

newEntry = oldEntry.copy()

simhue replied:
because then I have to write a new function that creates a deep copy of oldEntry, which gets used only once (so far), it means more code to maintain and touching additional pacakges, by calling getPatientById twice, I still get a separate object with "just" a single read operation on the database.

akomii replied:
I still dont get it... It a PUT, its called `updateEntry`, but all it does it update the comment?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+
+        newEntry = study.updatePatient(oldEntry, newEntry);
+
+        return Response.created(buildEntryLocation(newEntry)).entity(newEntry).build();
```
You return ok(), not created(). You did not create a new Entry....

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+                                                                @PathParam("reference") PatientReference ref,
+                                                                @PathParam("root") String root,
+                                                                PatientEntriesRequestDTO entries) throws IOException {
+        Study study = this.getStudy(id);
```
What if study not found?

simhue replied:
getStudy throws a NotFoundException

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/PatientEntriesResponseDTO.java:
```diff
@@ -0,0 +1,15 @@
+import org.aktin.dwh.optinout.PatientMasterData;
+
+@Getter @Setter
+public class PatientEntriesResponseDTO {
```
Shouldnt this be called [...]Entry[...]DTO instead of Entries?

simhue replied:
I named it like that because it was the response for PatientEntriesRequestDTO, but you are right

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+     * @throws IOException
+     */
+    @GET
+    public List<PatientEntry> getEntries() throws IOException {
```
Is this method even necessary? For what?

simhue replied:
it is definitely not used in the frontend, going to remove it

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+     * @return Response with status 'ok' and the list of entries
+     * @throws IOException
+     */
+    @Path("{studyId}")
```
Better studies/{studyId}. Also for all other Paths

simhue replied:
I'd rather wait with these changes until I am continuing with the tests, I am going to write a test case where the front end calls each endpoint, that would make any route changes easier to check

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+
+        if (study.getSicGeneration() == SICGeneration.AutoAndManual && (entry.sic == null || entry.sic.isEmpty())) {
+            entry.sic = study.generateSIC();
+        }
```
Such logic belongs in the service, not the endpoint

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+            validatedEntries.add(foundEntry);
+
+            if (!entries.generateSic) {
+                val sic = entries.sics.get(i);
```
What happens in case of an index drift?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+    private boolean validateSic(PatientEntriesRequestDTO entries, Study study, PatientEntriesResponseDTO foundEntry, String sic) throws IOException {
+        if (sic != null) {
+            if (entries.sics.stream().filter(s -> Objects.equals(s, sic)).count() > 1) {
+                foundEntry.setEntryValidation(EntryValidation.DUPLICATE_SIC);
```
Is this processed in the later steps somehow?

simhue replied:
Yes, in the frontend

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+
+        for (int i = 0; i < entries.extensions.size(); i++) {
+            val extension = entries.extensions.get(i);
+            val sic = entries.sics.get(i);
```
Again, the possibility of an index drift

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+            return true;
+        }
+
+        encounters.sort(Comparator.comparing(PatientEncounter::getStartDate));
```
Sort ascending and take first meaning the oldest entry?

akomii reviewed (commented):
Review WIP. Only Java code done so far. Project not buildable via mvn:

```
[ERROR] ▲ [WARNING] The glob pattern import("./**/*.entry.js*") did not match any files [empty-glob]
[ERROR] 
[ERROR]     node_modules/@stencil/core/internal/client/index.js:169:2:
[ERROR]       169 │   `./${bundleId}.entry.js${BUILD5.hotModuleReplacement && hmrVers...
[ERROR]           ╵   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ERROR] 
[ERROR] 
[ERROR] ✘ [ERROR] Could not resolve "semantic/dist/semantic.min.css"
[ERROR] 
[ERROR]     angular:styles/global:styles:2:8:
[ERROR]       2 │ @import 'semantic/dist/semantic.min.css';
[ERROR]         ╵         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ERROR] 
[ERROR]   You can mark the path "semantic/dist/semantic.min.css" as external to exclude it from the bundle, which will remove this error and leave the unresolved path in the bundle.
[ERROR] 
[ERROR] 
[ERROR] ✘ [ERROR] Could not resolve "semantic/dist/semantic.js" [plugin angular-script-global]
[ERROR] 
[ERROR] 
```

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+    public Response getEntriesByStudy(@PathParam("studyId") String id) throws IOException {
+        Study s = this.getStudy(id);
+        List<PatientEntry> list = new ArrayList<>();
+        list.addAll(s.allPatients());
```
s.allPatients() already returns List<? extends PatientEntry>

In getStudies() you do a cast, but here you are filling up a new list?

simhue replied:
Not my original code, will change to a cast nevertheless

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,492 @@
+    public PatientEntry getEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref, @PathParam("root") String root,
+                                 @PathParam("extension") String ext) throws IOException {
+        Study study = this.getStudy(id);
+        return study.getPatientByID(ref, root, ext);
```
This logic should be in the service. 

So you are calling in StudyManagerImpl an SQL Query to get all Studies
Then you loop to these studies to get a specific study
Then you loop through the found study to get a specific patient

Why not just add the method to StudyManager and implement a SQL Query that does this?

simhue replied:
Because then I do not need to touch two additional packages (dwh-api and dwh-query) and the computational impact should be negligible - correct me if I'm wrong

simhue committed:
reinsert accidentally deleted line

simhue commented:
> Review WIP. Only Java code done so far. Project not buildable via mvn:
> 
> ```
> [ERROR] ▲ [WARNING] The glob pattern import("./**/*.entry.js*") did not match any files [empty-glob]
> [ERROR] 
> [ERROR]     node_modules/@stencil/core/internal/client/index.js:169:2:
> [ERROR]       169 │   `./${bundleId}.entry.js${BUILD5.hotModuleReplacement && hmrVers...
> [ERROR]           ╵   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> [ERROR] 
> [ERROR] 
> [ERROR] ✘ [ERROR] Could not resolve "semantic/dist/semantic.min.css"
> [ERROR] 
> [ERROR]     angular:styles/global:styles:2:8:
> [ERROR]       2 │ @import 'semantic/dist/semantic.min.css';
> [ERROR]         ╵         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> [ERROR] 
> [ERROR]   You can mark the path "semantic/dist/semantic.min.css" as external to exclude it from the bundle, which will remove this error and leave the unresolved path in the bundle.
> [ERROR] 
> [ERROR] 
> [ERROR] ✘ [ERROR] Could not resolve "semantic/dist/semantic.js" [plugin angular-script-global]
> [ERROR] 
> [ERROR] 
> ```

During a merge a line got deleted by accident. I pushed a fix.

simhue committed:
https://github.com/aktin/dwh-admin/issues/85: use map instead of two lists
https://github.com/aktin/dwh-admin/pull/85#discussion_r2465779429

simhue committed:
#85: remove obsolote method

simhue committed:
#85: fix method description

simhue committed:
#85: rename PatientEntriesResponseDTO.java to PatientEntryResponseDTO.java

simhue committed:
#85: add new class

simhue committed:
#85: fix sort order

simhue committed:
#85: rework http error handling, no more sending human readable text as error responses

simhue committed:
#85: delete old studyManager folder and sources

simhue committed:
#85: add missing javadoc

simhue committed:
#85: change all endpoint return types to Response

simhue committed:
#85: remove unnecessary code

simhue committed:
#85: remove unused imports

simhue committed:
#85: remove provider

simhue committed:
#85: use js veriable to store preferences instead of session storage

simhue committed:
#85: change created response to ok

simhue committed:
#85: move sic generation from endpoint to service

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,461 @@
+    @Path("{studyId}/{sic}")
+    @GET
+    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
+    public PatientEntry getEntryBySic(@PathParam("studyId") String id, @PathParam("sic") String sic) throws IOException {
```
Endpoints should never throw an Exception... Where would it go?

simhue replied:
see comment above

rwm commented on admin-gui/pom.xml:
```diff
@@ -251,7 +263,16 @@
                 <version>3.1.0</version>
                 <configuration>
-                    <warSourceExcludes>node_modules/**,**/*.ts,**/*.js.map,.idea/**,.vscode/**</warSourceExcludes>
+                    <warSourceExcludes>node_modules/**,.angular/**,*.json,*.md,.browserslistrc,.vscode/**,.idea/**,dist/**,public/**,src/**</warSourceExcludes>
```
das soll ja den sourcecode enthalten und nur die kompilierten Teile auslassen. Warum steht hier public/ und src/ als ausnahmen drin?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,461 @@
+        pat = study.getPatientBySIC(entry.sic);
+        if (pat != null) {
+            throw OptInError.buildError(Status.CONFLICT, OptInErrorType.SIC_ALREADY_EXISTS, "Cannot create entry, SIC {0} already exists", entry.sic);
+        }
```
Again, an Endpoint is machine readable. Instead of creating a new Pojo to differentiate existing patient and sics you should use:

```java
throw new ClientErrorException(Status.CONFLICT);
```

Alternatively:
```java
throw new ClientErrorException(
    Response.status(Status.CONFLICT)
        .type(MediaType.APPLICATION_JSON)
        .entity("{detail: OptInErrorType.SIC_ALREADY_EXISTS}")
        .build());
```

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,461 @@
+    @Path("{studyId}/{reference}/{root}{p:/?}{extension:.*}")
+    @GET
+    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
+    public Response getEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref, @PathParam("root") String root,
```
Endpoints should never throw an IOException... Where would it go?

simhue replied:
Not my original code, we should adress that when the code base gets redesigned eventually

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,461 @@
+    public Response getEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref, @PathParam("root") String root,
+                                 @PathParam("extension") String ext) throws IOException {
+        Study study = this.getStudy(id);
+        return Response.ok(study.getPatientByID(ref, root, ext)).build();
```
If study not found, 200 is returned with an empty body

it should be a 404

simhue replied:
getStudy throws a not found exception when empty, but not getPatientByID, I'm going to add it

rwm commented on admin-gui/src/main/webapp/.gitignore:
```diff
@@ -5,3 +5,4 @@ npm-debug.log
 debug.log
 *.js
 *.js.map
+semantic
```
Hier fehlen kommentare was semantic ist und warum das ausgenommen wird.

rwm commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/ValidationErrorType.java:
```diff
@@ -0,0 +1,22 @@
+    ENCOUNTERS_NOT_FOUND, // No encounters in db found
+
+    // Processing States
+    VALID,
```
was für ein Fehler ist VALID???

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,461 @@
+     * @return list of studies
+     * @throws IOException
+     */
+    @Path("studies")
```
Better to just remove the Path annoation for consistency

GET optin/studies -> all studies
GET optin/2 -> all entries of study with id 2??

simhue replied:
or change the "optin" to "studies" because optin/2 -> what is an optin with the id 2?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,461 @@
+
+        PatientEntry oldEntry = study.getPatientByID(ref, root, ext);
+        if (oldEntry == null) {
+            throw OptInError.buildError(Status.NOT_FOUND, OptInErrorType.PATIENT_NOT_FOUND, "Patient not found");
```
Why do you need OptInError for this? Why not just:

```java
throw new ClientErrorException(Status.NOT_FOUND);
```

rwm commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/EntryDTO.java:
```diff
@@ -0,0 +1,13 @@
+@XmlAccessorType(XmlAccessType.FIELD)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class EntryDTO {
+	public String extension;
```
Javadoc fehlt komplett. Was ist "extension"? Variablenname schlecht gewählt.

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,461 @@
+            throw OptInError.buildError(Status.NOT_FOUND, OptInErrorType.PATIENT_NOT_FOUND, "Patient not found");
+        }
+        pat.delete(security.getUserPrincipal().getName());
+        return Response.ok().build();
```
Commonly, NO_CONTENT / 204 is returned if an DELETE is successful

simhue committed:
#85: add NotFoundException when PatientEntry is not found

simhue committed:
#89: refactor endpoint logic

simhue committed:
#89: add missing classes

akomii commented:
DWH is buildable, but not accessible:

<img width="2386" height="370" alt="image" src="https://github.com/user-attachments/assets/9e91831f-03ba-4fed-a91c-df0f401302b6" />

akomii commented:
`/optin` path is missing in `OptInEndpoint` and does not work:

<img width="1845" height="346" alt="image" src="https://github.com/user-attachments/assets/b1352786-da80-4fd9-80e7-66553652cf84" />

simhue committed:
#89: add missing rest endpoint annotations

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -70,40 +100,23 @@ public Response getEntriesByStudy(@PathParam("studyId") String id) throws IOExce
      */
-    @Path("{studyId}/{reference}/{root}{p:/?}{extension:.*}")
+    @Path("{studyId}/patients/{reference}/{extension}")
     @GET
```
<img width="1858" height="277" alt="Image" src="https://github.com/user-attachments/assets/ba6c34e8-9317-4e64-aaf0-7f24fbaa7b2d" />

Is this an error? The reference seems correct if I compare it with the business logic...

simhue replied:
This is a valid error

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -41,27 +56,42 @@ public class OptInEndpoint {
-
-        return Response.ok(s.allPatients()).build();
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getPatientsOfStudy(@PathParam("id") String id) {
```
<img width="1858" height="277" alt="Image" src="https://github.com/user-attachments/assets/157054e8-a0b0-41f8-bfa0-66f335e8b0a2" />

The response should be in a self-describing machine-readable structure. See JSON RFC 7807 for an example

simhue replied:
changed it to a model `{detail: "PATIENT_NOT_FOUND"}`.

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -41,27 +56,42 @@ public class OptInEndpoint {
+        try {
+            patients = patientService.getAllPatientsOfStudy(id);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Where does the RuntimeExecption go? The endpoint should be the final interface between business logic and outer world. Everything, that happens in the logic should be returned somehow as a response to the outer world.

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -112,326 +125,168 @@ public Response getEntryBySic(@PathParam("studyId") String id, @PathParam("sic")
-    @GET
-    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
+    @Path("patients/{reference}/encounters")
+    @POST
```
Why is this a POST? The method seems to be a getter

simhue replied:
Because of the parameter extensions, that list might be a long one. When using url parameters like you would with GET requests, that url could exceed URL max character limit. A common solution for that problem is to use a POST request and send parameters as body.

akomii replied:
Do you have an example for the payload? It seems like this method would actually be a search query. That is not clear from the url path and the method name

simhue replied:
`["4248733638","6790601213","3791899678","6451634779","7426931319","5977600398","8051844266","2381268180","7815303265","5647156122","5315103445","3694763933","9889845181","5745989379","5088635979","8763067297","6256782576","2708810975","5702679684","4609684415"]`

simhue replied:
"It seems like this method would actually be a search query. That is not clear from the url path and the method name" Do you have any suggestions for the url? I renamed the method for clarity

akomii replied:
If you want to make it clean, it would be something like:

public Response searchEncounters(EncounterSearchRequest request)


POST /encounters/search

{
  "patientReferences": "patient",
  "extensions": ["424", "425", "426"]
}

akomii replied:
I somehow cant find the patient via this method:

### Successful
<img width="1357" height="438" alt="Image" src="https://github.com/user-attachments/assets/3fcb18af-30d0-4f4d-bdf9-25816e72a047" />

### Not successful
<img width="1357" height="438" alt="Image" src="https://github.com/user-attachments/assets/509e9de3-5e43-4ea1-a37e-b91872949392" />

akomii replied:
Still empty response:

<img width="1765" height="453" alt="Image" src="https://github.com/user-attachments/assets/75171784-5be3-4e22-afcd-0c5b49274d6c" />

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -112,326 +125,168 @@ public Response getEntryBySic(@PathParam("studyId") String id, @PathParam("sic")
-    @Path("encounter/{reference}/{root}{p:/?}{extension:.*}")
-    @GET
-    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
+    @Path("patients/{reference}/encounters")
```
`/studies/patients/` reads a bit weird as an URI. Wouldnt it make sense to split it up into an `studies` endpoint and an `patients` endpoint?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -112,326 +125,168 @@ public Response getEntryBySic(@PathParam("studyId") String id, @PathParam("sic")
-    @Secured
-    @Path("{studyId}/{reference}/{root}{p:/?}{extension:.*}")
+    @Path("{studyId}/patients/{reference}/{extension}")
     @PUT
```
A PUT should be idempotent replacement. Here we have create-or-update logic with existence checks, meaning POST and PATCH behaviour

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -41,27 +56,42 @@ public class OptInEndpoint {
+            throw ErrorUtils.buildError(Status.NOT_FOUND, OptInErrorType.UNKNOWN, "An error occurred while retrieving studies:");
+        }
+        if(studies == null || studies.isEmpty()) {
+            throw ErrorUtils.buildError(Status.NOT_FOUND, OptInErrorType.STUDIES_NOT_FOUND, "No studies found");
```
An empty collection is still a valid response, so it must return 200 either way

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -70,40 +100,23 @@ public Response getEntriesByStudy(@PathParam("studyId") String id) throws IOExce
-        PatientEntry patientEntry = study.getPatientBySIC(sic);
         if(patientEntry == null) {
-            throw OptInError.buildError(Status.NOT_FOUND, OptInErrorType.PATIENT_NOT_FOUND, "Patient not found");
+            throw ErrorUtils.buildError(Status.NOT_FOUND, OptInErrorType.PATIENT_NOT_FOUND, "Patient not found");
```
This 404 is correct, because the patient resource may be absent

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -112,326 +125,168 @@ public Response getEntryBySic(@PathParam("studyId") String id, @PathParam("sic")
+            throw new RuntimeException(e);
+        }
+        if(encounters == null || encounters.isEmpty()) {
+            throw ErrorUtils.buildError(Status.NOT_FOUND, OptInErrorType.ENCOUNTERS_NOT_FOUND, "Encounters not found");
```
Again, an empty collection is still a valid resposne

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -112,326 +125,168 @@ public Response getEntryBySic(@PathParam("studyId") String id, @PathParam("sic")
-        if (pat == null) {
-            throw OptInError.buildError(Status.NOT_FOUND, OptInErrorType.PATIENT_NOT_FOUND, "Patient not found");
+    public Response deleteEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref,
+                                @PathParam("extension") String ext) {
```
<img width="1752" height="391" alt="Image" src="https://github.com/user-attachments/assets/57b97c27-6450-49f6-a142-336a1956bae7" />

Deleting an non-existing patient yields an ServerError. A DELETE should be idempotent

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -112,326 +125,168 @@ public Response getEntryBySic(@PathParam("studyId") String id, @PathParam("sic")
+    @PUT
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response createEntries(@PathParam("studyId") String id,
```
Why do we have `saveEntry()` and here `createEntries()` ?

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -112,326 +125,168 @@ public Response getEntryBySic(@PathParam("studyId") String id, @PathParam("sic")
-        newEntry.setComment(entry.comment);
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response saveEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref,
```
<img width="1752" height="391" alt="Image" src="https://github.com/user-attachments/assets/7254a9d5-b019-4210-b071-f044bff3a4fd" />

Malformed payload yields an ServerError instead of BAD REQUEST

akomii reviewed (changes requested):
See comments attached, in summary:

- Empty collections should be returned as 200/`[ ]`
- Read operations implemented as POST
- Wrong PUT semantics
- Exceptions are not properly formatted

simhue committed:
#89:
- improve deserialisation validation
- move files to proper package
- improve error handling
- change method name(s) and http methods used

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -103,108 +96,89 @@ public Response getPatientsOfStudy(@PathParam("id") String id) {
-                                @PathParam("extension") String ext, PatientEntryRequestDTO entry) {
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response updateEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref,
```
PUT does not work (no changes but comment)

<img width="1216" height="475" alt="Image" src="https://github.com/user-attachments/assets/43f3ba67-978f-4526-adc6-d07d72234b2c" />

simhue committed:
#89: - added comment

simhue committed:
#89: - rename methods

simhue committed:
#89: - change http method

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -151,19 +152,19 @@ public Response getMasterData(@PathParam("reference") PatientReference ref,
     @Consumes(MediaType.APPLICATION_JSON)
     @Produces(MediaType.APPLICATION_JSON)
-    public Response updateEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref,
+    public Response partlyUpdateEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref,
```
partly is ambiguous. What are you exactly updating? How I understood, only the comments?

simhue replied:
I kept it ambiguous for possible later changes where comment isn't the only field being updated, but I will change the name

simhue committed:
#89: - change method name and comment

simhue committed:
#89: - change url paths and parameter signatures

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,289 @@
+    @Path("{id}/patients")
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getPatientsOfStudy(@PathParam("id") String id) {
```
nitpick: Should throw an error if study does not exist

simhue replied:
should be solved in dwh-query then?

akomii replied:
No in this iteration at least

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -1,218 +1,289 @@
+     * @param ref  patient reference
+     * @param extensions  extensions
+     * @return list of patient encounters
+     */
```
nitpick: Dont forget to update docs too

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
```diff
@@ -119,51 +121,57 @@ public Response getPatient(@PathParam("studyId") String id,
     @Consumes(MediaType.APPLICATION_JSON)
     @Produces(MediaType.APPLICATION_JSON)
-    public Response updateEntry(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref,
+    public Response updateEntryComment(@PathParam("studyId") String id, @PathParam("reference") PatientReference ref,
```
Why such a big DTO if you only update the comment?

simhue replied:
Ultimately it is only necessary to write the audit log, I am going to open an issue in dwh-query regarding this

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
Is this correct, that this is a PUT?

simhue replied:
No, I changed it to POST

akomii commented on admin-gui/src/main/java/org/aktin/dwh/admin/optin/OptInEndpoint.java:
What is wrong here? The log is not helpful at all

<img width="2217" height="621" alt="Image" src="https://github.com/user-attachments/assets/fe1df1a0-7fee-4ba6-bde7-dfd105ac6928" />

<img width="1495" height="68" alt="Image" src="https://github.com/user-attachments/assets/fb9cdc2e-500c-419e-88bc-511e1ba4bc92" />

simhue replied:
Improved logging by adding stacktrace

simhue committed:
#89: improve loggging

simhue committed:
#89: improve logging

simhue committed:
#89: change comment

simhue committed:
#89: change http method

akomii committed:
Merge branch 'develop' into feat/consent-manager-20

Signed-off-by: Alexander Kombeiz <30117682+akomii@users.noreply.github.com>

akomii commented:
### Context

1. Given this patient (P1000):

<img width="846" height="561" alt="Screenshot from 2026-03-02 14-29-02" src="https://github.com/user-attachments/assets/e15dda07-6c2b-445b-90bc-88e163ed3df7" />

2. We find him in the DWH as patient_num=3:

<img width="2093" height="116" alt="image" src="https://github.com/user-attachments/assets/3d386d68-4110-4984-96bf-d246720b814a" />

3. Accordingly, encounter_num=3:

<img width="1648" height="116" alt="image" src="https://github.com/user-attachments/assets/e9946e8c-898d-4e4b-b9f2-7a539cc414f8" />

4. Here we see the ide hashes:

<img width="1872" height="122" alt="image" src="https://github.com/user-attachments/assets/bad34496-48c5-4dcd-89f6-04d576b70c6d" />

5. The POST requests for encounter and masterdata:

<img width="1617" height="360" alt="image" src="https://github.com/user-attachments/assets/3a62f703-eb7a-4f74-aae7-4a26be9ae2a9" />

<img width="1617" height="360" alt="image" src="https://github.com/user-attachments/assets/2b5e7c4f-b99a-4d34-9a03-a3014a5dc67b" />

### Problems
- `idEnc` is wrong (`pat_ide` is used)
- `zip` is `null` (should be 80636)
- `start_date` (Instant) is 1/17/2015, 16:03:00
- `end_date` (Instant) is 1/17/2015, 17:35:00

simhue committed:
#89: rename encounter to encounterperiods and idEnc to ide

simhue committed:
Merge remote-tracking branch 'origin/feat/consent-manager-20' into feat/consent-manager-20

akomii committed:
fix: removed leading slashes for "encounterperiods" and "masterdata"

simhue committed:
Implementation of Consent Manager 2.0 frontend (#100)

Signed-off-by: Simon H. <huening.simon@gmail.com>
